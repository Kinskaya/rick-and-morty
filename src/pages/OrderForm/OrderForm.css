:root {
    --color-white: rgba(255,255,255,0.3);
}

.order-form {
    display: flex;
    flex-direction: column;
    margin: 0 auto;
    max-width: 500px;
    height: auto;
    border-radius: 20px;
    background-color: var(--color-white);
    box-shadow: 20px 20px 40px -6px rgba(0,0,0,0.2);
    transition: all 0.2s ease-in-out;

}
.toggle-switch {
    position: relative;
    display: inline-block;
    width: 50px;
    height: 25px;
    margin-left: 10px;
}
.toggle-switch input[type="checkbox"] {
    display: none;
}
.toggle-switch .switch {
    position: absolute;
    cursor: pointer;
    background-color: #ccc;
    border-radius: 25px;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    transition: background-color 0.2s ease;
}
.toggle-switch .switch::before {
    position: absolute;
    content: "";
    left: 2px;
    top: 2px;
    width: 21px;
    height: 21px;
    background-color: #aaa;
    border-radius: 50%;
    transition: transform 0.3s ease;
}
.toggle-switch input[type="checkbox"]:checked + .switch::before {
    transform: translateX(25px);
    background-color: #6699cc;
}
.toggle-switch input[type="checkbox"]:checked + .switch {
    background-color: #414191;
    padding-left: 4px;
}

::placeholder,
select {
    font-family: Montserrat, sans-serif;
    font-weight: 400;
}

.form-control {
    display: flex;
    margin-bottom: 10px;
}

.form-control label,
.toggle-desc {
    font-weight: 500;
    opacity: 0.7;
    font-size: 1rem;
    margin-top: 0;
    text-shadow: 2px 2px 4px rgba(0,0,0,0.2);
}

input,
select {
    margin-left: 10px;
}

form input[type=submit] {
    margin-top: 10px;
    width: 150px;
    font-size: 1rem;
    background: transparent;
    padding: 0.8em;
    margin-bottom: 1em;
    border: none;
    border-left: 1px solid rgba(255, 255, 255, 0.3);
    border-top: 1px solid rgba(255, 255, 255, 0.3);
    border-radius: 50px;
    box-shadow: 4px 4px 60px rgba(0, 0, 0, 0.2);
    transition: all 0.2s ease-in-out;
    text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.2);
}

.errors {
    color: #c60f38;
    font-size: 0.8em;
    margin: -10px 0 10px 0
}

.errors.name-info {
    margin-left: 87px;
}

.errors.date-info,
.errors.place-info {
    margin-left: 108px;
}

.errors.file-info {
    margin-right: 54px;
}
