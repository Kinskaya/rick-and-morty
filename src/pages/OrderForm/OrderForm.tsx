import { useEffect } from 'react';
import { Path, SubmitHandler, useForm, UseFormRegister } from 'react-hook-form';
import { useDispatch, useSelector } from 'react-redux';

import NameInput from './components/NameInput/NameInput';
import DateInput from './components/DateInput/DateInput';
import PlaceSelect from './components/PlaceSelect/PlaceSelect';
import Reminder from './components/Reminder/Reminder';
import Toggle from './components/Toggle/Toggle';
import FileSelect from './components/FileSelect/FileSelect';
import OrderList from '../../components/OrderList/OrderList';

import { AppDispatch, RootState } from '../../redux/store';
import { onInput } from '../../redux/ordersSlice';

import './OrderForm.css';

export type TOrderForm = {
  nameInput?: string;
  placeSelect?: string;
  reminder?: boolean;
  isToggled?: boolean;
  dateInput?: string;
  selectedFile?: File;
};

export type TFormInput = {
  nameInput?: string;
  placeSelect?: string;
  reminder?: boolean;
  isToggled?: boolean;
  dateInput?: string;
  selectedFile?: File[];
};

export type InputProps = {
  label: Path<TFormInput>;
  register: UseFormRegister<TFormInput>;
  required?: boolean;
};

const OrderForm = () => {
  const { ...order } = useSelector<RootState>((state) => state.orders.order) as TFormInput;

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isSubmitSuccessful },
  } = useForm<TFormInput>({
    defaultValues: { ...order },
  });

  const dispatch = useDispatch<AppDispatch>();
  const orders = useSelector<RootState>((state) => state.orders.orders) as TOrderForm[];

  const onSubmit: SubmitHandler<TFormInput> = (data: TFormInput): void => {
    const order = {
      nameInput: data.nameInput,
      placeSelect: data.placeSelect,
      reminder: data.reminder,
      isToggled: data.isToggled,
      dateInput: data.dateInput,
      selectedFile: data.selectedFile?.[0],
    };

    dispatch(onInput(order));
    alert('Your order is saved successfully');
  };

  useEffect(() => {
    if (isSubmitSuccessful) {
      reset({
        nameInput: '',
        placeSelect: '',
        reminder: false,
        isToggled: false,
        dateInput: '',
        selectedFile: [],
      });
    }
  }, [isSubmitSuccessful, reset]);

  return (
    <>
      <form className="order-form" onSubmit={handleSubmit(onSubmit)}>
        <h3>Order form</h3>
        <NameInput label={'nameInput'} register={register} required />
        {errors['nameInput'] && <p className="errors name-info">{errors['nameInput']?.message}</p>}
        <DateInput label={'dateInput'} register={register} required />
        {errors['dateInput'] && <p className="errors date-info">{errors['dateInput']?.message}</p>}
        <PlaceSelect label={'placeSelect'} register={register} required />
        {errors['placeSelect'] && (
          <p className="errors place-info">{errors['placeSelect']?.message}</p>
        )}
        <Reminder label={'reminder'} register={register} />
        <Toggle label={'isToggled'} register={register} />
        <FileSelect label={'selectedFile'} register={register} required />
        {errors['selectedFile'] && <p className="errors file-info">{'File can not be empty!'}</p>}
        <input type="submit" value="Submit" disabled={!isDirty} />
      </form>
      <OrderList orders={orders} />
    </>
  );
};
export default OrderForm;
