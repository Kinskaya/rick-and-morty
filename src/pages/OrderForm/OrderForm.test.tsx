import OrderForm from './OrderForm';
import { render } from '@testing-library/react';

describe('OrderForm component', () => {
  it('render correctly', () => {
    const { asFragment } = render(<OrderForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
