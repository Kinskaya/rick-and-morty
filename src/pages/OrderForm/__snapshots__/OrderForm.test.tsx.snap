// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderForm component render correctly 1`] = `
<DocumentFragment>
  <form
    class="order-form"
  >
    <h3>
      Order form
    </h3>
    <div
      class="form-control"
    >
      <label>
        Add name
      </label>
      <input
        name="nameInput"
        placeholder="Your name"
        type="text"
      />
    </div>
    <div
      class="form-control"
    >
      <label>
        Delivery date
      </label>
      <input
        min="2022-04-19"
        name="dateInput"
        type="date"
      />
    </div>
    <div
      class="form-control"
    >
      <label>
        Delivery place
      </label>
      <select
        name="placeSelect"
      >
        <option
          value=""
        >
          Select place
        </option>
        <option
          value="Minsk"
        >
          Minsk
        </option>
        <option
          value="Gomel"
        >
          Gomel
        </option>
        <option
          value="Mogilev"
        >
          Mogilev
        </option>
        <option
          value="Brest"
        >
          Brest
        </option>
        <option
          value="Grodno"
        >
          Grodno
        </option>
      </select>
    </div>
    <div
      class="form-control"
    >
      <label>
        Contact before delivery
      </label>
      <input
        name="reminder"
        type="checkbox"
      />
    </div>
    <div
      class="form-control"
    >
      <span
        class="toggle-desc"
      >
        I want to receive notifications about promo
      </span>
      <label
        class="toggle-switch"
      >
        <input
          name="isToggled"
          type="checkbox"
        />
        <span
          class="switch"
        >
          Yes
        </span>
      </label>
    </div>
    <div
      class="form-control"
    >
      <label>
        Upload file
        <input
          accept=".jpg, .jpeg, .png"
          name="selectedFile"
          type="file"
        />
      </label>
    </div>
    <input
      disabled=""
      type="submit"
      value="Submit"
    />
  </form>
  <div
    class="card-container"
  />
</DocumentFragment>
`;
